pkg/drivers/krunkit/krunkit.go

```diff
@@ -0,0 +1,562 @@
+		return err
+	}
+
+	if d.ExtraDisks > 0 {
+		log.Info("Creating extra disk images...")
+		for i := 0; i < d.ExtraDisks; i++ {
+			path := pkgdrivers.ExtraDiskPath(d.BaseDriver, i)
+			if err := pkgdrivers.CreateRawDisk(path, d.DiskSize); err != nil {
+				return err
+			}
+		}
+	}
+
+	log.Info("Starting krunkit VM...")
+	return d.Start()
+}
+
+func (d *Driver) Start() error {
+	socketPath := d.VmnetHelper.SocketPath()
+	if err := d.VmnetHelper.Start(socketPath); err != nil {
+		return err
+	}
+
+	d.MACAddress = d.VmnetHelper.GetMACAddress()
+
+	if err := d.startkrunkit(socketPath); err != nil {
+		return err
+	}
+
+	if err := d.setupIP(d.MACAddress); err != nil {
+		return err
+	}
+
+	log.Infof("Waiting for VM to start (ssh -p %d docker@%s)...", d.SSHPort, d.IPAddress)
+
+	return WaitForTCPWithDelay(fmt.Sprintf("%s:%d", d.IPAddress, d.SSHPort), time.Second)
+}
+
+// startkrunkit starts the krunkit child process.
+func (d *Driver) startkrunkit(socketPath string) error {
+	var args = []string{
+		"--memory", fmt.Sprintf("%d", d.Memory),
+		"--cpus", fmt.Sprintf("%d", d.CPU),
+		"--restful-uri", d.restfulURI(),
+		"--device", fmt.Sprintf("virtio-net,unixSocketPath=%s,mac=%s", socketPath, d.MACAddress),
+		"--device", fmt.Sprintf("virtio-serial,logFilePath=%s", d.serialPath()),
+		"--krun-log-level", logLevelInfo,
+
+		// The first device is the boot disk.
+		"--device", fmt.Sprintf("virtio-blk,path=%s", d.isoPath()),
+		"--device", fmt.Sprintf("virtio-blk,path=%s", d.diskPath()),
+	}
+
+	for i := 0; i < d.ExtraDisks; i++ {
+		args = append(args,
+			"--device", fmt.Sprintf("virtio-blk,path=%s", pkgdrivers.ExtraDiskPath(d.BaseDriver, i)))
+	}
+
+	log.Debugf("executing: krunkit %s", strings.Join(args, " "))
+	cmd := exec.Command(driverName, args...)
+
+	// Create krunkit in a new process group, so minikube caller can use killpg
+	// to terminate the entire process group without harming the krunkit process.
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	logfile, err := d.openLogfile()
+	if err != nil {
+		return fmt.Errorf("failed to open krunkit logfile: %w", err)
+	}
+	defer logfile.Close()
+	cmd.Stderr = logfile
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	return process.WritePidfile(d.pidfilePath(), cmd.Process.Pid)
+}
+
+func (d *Driver) openLogfile() (*os.File, error) {
+	logfile := d.ResolveStorePath(logFileName)
+	return os.OpenFile(logfile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+}
+
+func (d *Driver) setupIP(mac string) error {
+	var err error
+	getIP := func() error {
+		d.IPAddress, err = pkgdrivers.GetIPAddressByMACAddress(mac)
+		if err != nil {
+			return errors.Wrap(err, "failed to get IP address")
+		}
+		return nil
+	}
+	// Implement a retry loop because IP address isn't added to dhcp leases file immediately
+	for i := 0; i < 60; i++ {
+		log.Debugf("Attempt %d", i)
+		err = getIP()
+		if err == nil {
+			break
+		}
+		time.Sleep(2 * time.Second)
```

Comment:
curious how did we come up with 2 seconds, why not 1 ?

Reply:
Maybe to limit cpu usage, reading and parsing /var/db/dhcpd_leases can be expensive. But this is pretty bad, causing unnecessary delays up to 2 seconds when waiting for ip.

But improving this is not in the scope of adding new driver, I'll open an issue for this.